style.md

```diff
@@ -0,0 +1,281 @@
+```julia
+# Bad!
+N = 5
+x = [] #Really bad since it is an Any vector!
+for i in 1:N
+    push!(x, 2.0 * i^2)
+end
+x
+
+# Better!
+x = ones(N)
+for i = 1:N
+    x[i] = 2.0 * i^2
+end
+x
+
+# Best!
+x = [2.0 * i^2 for i in 1:N]
+
+# or
+f(i) = 2.0 * i^2
+x = f.(1:5) #Use broadcasting
+```
+- Prefer `eachindex` to accessing the sizes of matrices.
+```julia
+x = [1.0, 2.0, 5.0, 2.1]
+y = similar(x)
+
+# BAD!
+for i in 1:length(x)
+    y[i] = x[i] + 2
+end
+y
+
+# BETTER!
+for i in eachindex(x)
+    y[i] = x[i] + 2
+end
+y
+
+# GOOD! #No way to preallocate y (although easier ways to write)
+x = rand(10)
+y = similar(x, 0) #Empty of same type as x
+for val in x
+    if val < 0.5
+        push!(y, val)
+    end
+end
+y
+```
+- **Use iterators directly** rather than accessing by index.
+```julia
+x = [1, 2, 3]
+
+# BAD!
+n = 0
+for i in eachindex(x)
+  n += x[i]^2
+end
+n
+
+# BETTER!
+n = 0
+for val in x
+  n += val^2
+end
+n
+
+# BEST!
+sum(xval -> xval^2, x) #i.e. transform each x and then reduce
+```
+
+- **Avoid `range` when possible** and use the `1.0:0.1:1.0` style notation, etc.  Getting rid of the `linspace` is a pain, but use ranges if the exact top doesn't matter.
+```julia
+Base.OneTo(5) # BAD!
+range(1, stop=5) #BAD!
+1:5 #GOOD!
+
+
+N = 10
+zmin = 0.0
+zstop = 1.0
+
+# Old linspace(zmin, zstop, N) doesn't work
+# ACCEPTABLE!
+range(zmin, stop = zstop, length = N)
+```    
+  - Furthermore, if you don't really care if if hits the `zstop` exactly and are willing to give a stepsize then, the following is the clearest
+```julia
+zmin = 0.0
+zstop = 1.0
+step = 0.1
+
+# GOOD! But..
+r = zmin:step:zstop
+
+# ...CAREFUL
+r = 0.0:0.22:1.0 # Note the end isn't a multiple of the step...
+@assert r == 0.0:0.22:0.88
+@assert maximum(r) == 0.88 #Use to get the maxium of the range, perhaps != 
+```
+
+
+
+## Dependencies
+
+- **Use external packages** whenever possible, and never rewrite code that is available in a well-maintained external package (even if it is imperfect)
+- The following packages can be used as a dependency without any concerns: `QuantEcon, Parameters, Optim, Roots, Expectations, NLsolve, DataFrames, LaTeXStrings, Plots, ....`
+- **Do** use `using` where possible (i.e. not `import`), and include the whole package as opposed to selecting only particular functions or types.
+- **Prefer** to keep packages used throughout the lecture at the top of the first block (e.g. `using LinearAlgebra, Parameters, QuantEcon`)  but packages used only in a single place should have the `using` local to that use.
+    - If `Plots` is only used lower down in the lecture, then try to have it local to that section to ensure faster loading time.
+- **Always seed random numbers** in order for automated testing to function.
+
+## Work in Progress Discussions
+1. When and how do we use `eachindex` with multidimensional arrays you want to iterate through?
+2. How best to stack arrays and unpack them for use with solvers/etc.?  
```

Comment:
That's very MATLAB-ish. I think most of Optim-land doesn't require it anymore.

Reply:
Alas, I am not sure that is true outside of `JuMP`.    I am running into it every day when doing the DAE problems, solving bellman equations with different discrete states, etc.

You will also see it if you collapse some of the dynamic programming problems here, such as the McCall model.